design/deploy-steps.md

```diff
@@ -0,0 +1,134 @@
+- A way to provide a link to the Ironic agent image per host.
+
+### Non-Goals
+
+- Expose the full functionality of Ironic deploy steps.
+- Building custom agent images.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a CoreOS user I want to use `coreos-installer` to write the CoreOS image
+without losing any functionality that Ironic offers (unlike the live ISO
+workflow).
+
+#### Story 2
+
+As an operator I want to execute site-specific actions during deployment. I'm
+ready to provide them as plugins to the Ironic agent inside my custom agent
+image.
+
+## Design Details
+
+Additions to `BareMetalHost`:
+
+- New field `AgentImageURL`. It will accept an optional link to a custom agent
+  ISO image.
+
+NOTE: custom kernel/initramfs are out of scope for this proposal.
+
+- New field `DeploySteps` with a list of deploy steps to apply. Each step
+  is a structure with fields `Name` (string) and `Args` (JSON).
+- New boolean field `SkipDefaultDeploySteps` defaulting to `false`. If `true`,
+  the execution of the default deployment procedure will be supressed.
+- The `Image.URL` field will become optional if `SkipDefaultDeploySteps` is
+  `true` since we don't know if the deploy method actually requires a separate
+  image.
+
+NOTE: `coreos-installer` can use the root filesystem of the ramdisk itself
+without downloading another image.
+
+### Implementation Details/Notes/Constraints
+
+Ironic accepts custom deploy steps in the provisioning API (see [gophercloud
+change][gc-deploy-steps]). Deploy steps have numeric priorities, which we will
+calculate automatically by starting with 75 and going down with each step (75,
+74, ...). The `Interface` field will always be set to `deploy`.
+
+NOTE: the value 75 is picked so that the execution starts after the image is
+written but before the bootloader is installed.
+
+When `SkipDefaultDeploySteps` is `true`, we will do the above plus will use the
+new [custom-agent deploy interface][custom-agent] so that the default steps do
+not apply. We will also not require `Image.URL`.
+
+[gc-deploy-steps]: https://github.com/gophercloud/gophercloud/pull/2120
+[custom-agent]: https://review.opendev.org/c/openstack/ironic/+/786033
+
+### Risks and Mitigations
+
+The new functionality will be easy to misuse, especially if you're not familiar
+with Ironic. We will need to write the documentation accordingly.
+
+### Work Items
+
+- Add support for per-host agent images (`AgentImageURL`).
+- Add support for accepting deploy steps on `BareMetalHost`.
+- Implement [custom-agent][custom-agent] in Ironic.
+- Add support for supressing the default deploy procedure.
+- Fix boot interface validation in Ironic that makes `image_source` always
+  required, even if the deploy interface does not use it.
+
+### Dependencies
+
+- [custom-agent deploy interface in Ironic][custom-agent]
+
+### Test Plan
+
+We can potentially add CI jobs with CoreOS, although it may require some
+OpenShift bits.
```

Comment:
A minimal test could deploy an FCOS image and ensure that it boots, without worrying about whether it can be a node.